library/modules/Units.cpp

```diff
@@ -1942,3 +1944,153 @@ int Units::getStressCategoryRaw(int32_t stress_level)
+    case unit_action_type::Job2:
+        return &action->data.job2.timer;
+        // could also patch the unit->job.current_job->completion_timer
+    case unit_action_type::PushObject:
+        return &action->data.pushobject.timer;
+    case unit_action_type::SuckBlood:
+        return &action->data.suckblood.timer;
+    case unit_action_type::Jump:
+    case unit_action_type::ReleaseTerrain:
+    case unit_action_type::Parry:
+    case unit_action_type::Block:
+    case unit_action_type::HoldItem:
+    case unit_action_type::ReleaseItem:
+    default:
+        break;
+    }
+    return nullptr;
+}
+
+void Units::subtractActionTimer(df::unit *unit, int amount, df::unit_action_type affectedActionType)
+{
+    CHECK_NULL_POINTER(unit);
+    for (auto action : unit->actions) {
+        if (affectedActionType != action->type) continue;
+        int32_t *timer = getActionTimerPointer(action);
+        if (timer != nullptr && *timer != 0) {
+            *timer = max(*timer - amount, 1);
+        }
+    }
+}
+
+void Units::subtractCategoryActionTimers(df::unit *unit, int amount, df::unit_action_type_group affectedActionTypes)
+{
+    CHECK_NULL_POINTER(unit);
+    for (auto action : unit->actions) {
+        auto list = ENUM_ATTR(unit_action_type, group, action->type);
+        for (size_t i = 0; i < list.size; i++) {
+            if (list.items[i] == affectedActionTypes) {
+                int32_t *timer = getActionTimerPointer(action);
+                if (timer != nullptr && *timer != 0) {
+                    *timer = max(*timer - amount, 1);
+                }
+                break;
+            }
+        }
+    }
+}
+
+void Units::multiplyActionTimer(df::unit *unit, float amount, df::unit_action_type affectedActionType)
+{
+    CHECK_NULL_POINTER(unit);
+    for (auto action : unit->actions) {
+        if (affectedActionType != action->type) continue;
+        int32_t *timer = getActionTimerPointer(action);
+        if (timer != nullptr && *timer != 0) {
+            double value = *timer;
+            value = max(value * amount, 1.0);
+            if (value > INT32_MAX) {
+                value = INT32_MAX;
+            }
+            *timer = value;
+        }
+    }
+}
+
+void Units::multiplyCategoryActionTimers(df::unit *unit, float amount, df::unit_action_type_group affectedActionTypes)
+{
+    CHECK_NULL_POINTER(unit);
+    for (auto action : unit->actions) {
+        auto list = ENUM_ATTR(unit_action_type, group, action->type);
+        for (size_t i = 0; i < list.size; i++) {
+            if (list.items[i] == affectedActionTypes) {
+                int32_t *timer = getActionTimerPointer(action);
+                if (timer != nullptr && *timer != 0) {
+                    double value = *timer;
+                    value = max(value * amount, 1.0);
+                    if (value > INT32_MAX) {
+                        value = INT32_MAX;
+                    }
+                    *timer = value;
+                }
```

Comment:
This is a copy of the code in multiplyActionTimer. You should factor this out so you only have to fix bugs in it once. Similarly for subtractCategoryActionTimers and setCategoryActionTimers.

Reply:
OK, done, will push